### PR TITLE
Qualified names for Lustre Identifiers

### PR DESCRIPTION
**Pull request by apoorvingle**
Features
-------

- AST identifiers are now qualified
- Modes are stored as a qualified identifier and not a list of strings
- ghost variables and constants of imported contracts can be accessed by a qualified manner.
- Contracts can be imported using `as` construct i.e. aliased imports.

### CODE REVIEW

File: src/utils/lib.mli [hunk 1]
```diff
@@ -63,6 +63,9 @@ val get : 'a option -> 'a
 (** Return the min between two optional floats. Return None if both floats are None. *)
 val min_option : float option -> float option -> float option
 
+(** Maps a function over an optional value *)
+val map_option: ('a -> 'b) -> 'a option -> 'b option 
+
```
Comment (daniel-larraz) [inline src/utils/lib.mli:68]: Mention this function is available in version 4.08 and later of the standard library:
https://caml.inria.fr/pub/docs/manual-ocaml/libref/Option.html

File: src/lustre/lustreLexer.mll [hunk 2]
```diff
@@ -334,7 +335,10 @@ let keyword_table = mk_hashtbl [
    C syntax: alphanumeric characters including the underscore, starting 
    with a letter or the underscore *)
 let id = ['a'-'z' 'A'-'Z' '_'] ['a'-'z' 'A'-'Z' '_' '0'-'9']*
-         
+
+let qual_id = id (':' ':' id)*
+(* a qualified id is an id or id::qual_id *)
+
```
Comment (daniel-larraz) [inline src/lustre/lustreLexer.mll:341]: A qualified identifier shouldn’t be a single token, but a grammar rule. This forces the parser to deconstruct the qualified identifier unnecessarily.

File: src/utils/lib.mli (hunk 1 above)
Comment (apoorvingle) [reply src/utils/lib.mli:68]: yup. but we are using 4.07 right? (:

Comment (daniel-larraz) [reply src/utils/lib.mli:68]: I meant please write a comment to future developers.

Comment (apoorvingle) [reply src/utils/lib.mli:68]: Ah, i misread your comment..

**Pull request closed without merging**